### ISSUE TITLE
Request after server closure is still served

### ISSUE THREAD

**Issue by tlienart** [JuliaWeb/HTTP.jl#405]
Long story short: I believe there are circumstances in which @async threads remain alive even after closing the server that is being listened on. This means that you can still access pages as if they are still being served while the status of the server is closed (6). This can cause problems when starting and closing a server multiple times in one Julia session. I present a MWE below that reproduces the problem + a possible fix.

Julia 1.2
HTTP.jl 0.8#master
Mac
latest Safari & Firefox

## MWE

**Edit**: Jonas provided a much simpler example [below](https://github.com/JuliaWeb/HTTP.jl/issues/405#issuecomment-479227654).

Here's a simple example with a single websocket:

**index.html**
```html
<!DOCTYPE html>
<html><body>Some text</body><html>
```

**script**
```julia
using HTTP
using Sockets
using Test

const JSWS = """
    <script type="text/javascript">
      var sock = new WebSocket("ws://" + location.host + location.pathname);
      sock.onmessage = function(msg){sock.close();};
    </script>
    """

server = Sockets.listen(8000)

listener = @async HTTP.listen(Sockets.localhost, 8000, server=server) do http::HTTP.Stream
    if HTTP.WebSockets.is_upgrade(http.message)
        println("upgrade")
        HTTP.setstatus(http, 101)
        HTTP.setheader(http, "Upgrade" => "websocket")
        HTTP.setheader(http, "Connection" => "Upgrade")
        key = HTTP.header(http, "Sec-WebSocket-Key")
        HTTP.setheader(http, "Sec-WebSocket-Accept" => HTTP.WebSockets.accept_hash(key))
        HTTP.startwrite(http)

        io = http.stream
        ws = HTTP.WebSockets.WebSocket(io; server=true)
        write(ws, "ping")
        println("pinged")
    elseif http.message.method == "GET" && http.message.target ∈ ("/", "/index.html")
        h = HTTP.Handlers.RequestHandlerFunction() do req::HTTP.Request
            content = read("index.html", String)
            content = replace(content, "</body>" => "$JSWS</body>")
            HTTP.Response(200, content)
         end
        HTTP.Handlers.handle(h, http)
    end
end
```

What this does:

* serves `index.html` after appending a short javascript which opens a websocket and closes it upon receiving a message. 

What is problematic:

_sequence of actions_

* open a browser at localhost, works fine, will write "upgrade" and "pinged" in the console
* in Julia:

```julia
close(server)
@test server.status == 6
sleep(0.05) # yield so that the async task can be done for sure
@test istaskdone(listener)
```

all this should pass fine

* get back to the browser and reload, **the page is still accessible** (until you kill Julia)

**edit**: the reason this is problematic is that it can cause weird bugs and errors when re-starting a server in the same Julia session. In any case it's not a good thing that stuff lives on by itself when the user should assume everything to be shut down.

## A possible explanation and fix

What I think may be causing an issue is lying in the `listenloop` logic and subsequent `@async` threads it launches here and there. These `@async` threads possibly **do not see when the server is closed** and therefore live on uninterrupted until julia is killed.

In order to try this, I modified the code as follows:

1. in `listenloop` pass `server` as the last argument to `handle_connection`
1. modify `handle_connection` to take `server` as last argument and check explicitly that the server is opened and if it isn't, close the connection:

```julia
while isopen(c)
    if isopen(server)
        handle_transaction(f, Transaction(c), server;
                           final_transaction=(count == reuse_limit))
        count += 1
    else
        close(c)
    end
end
```

Finally modify `handle_transaction` as well to take `server` as an argument and

```julia
@async try
    if isopen(server)
        f(http)
        closeread(http)
        closewrite(http)
    end
catch # ...
```

This may not be the most elegant way to solve the problem, however it does solve the issue.

## Next step

* I don't know HTTP.jl very well and my analysis may be completely wrong, in that case I apologise for the spam but would welcome explanations to understand what's going on
* If I'm more or less right, I'll be happy to provide a PR for the above suggested fix or anything better that is suggested.

Cheers

**Comment by asprionj** [comment]
Here's a simpler example, I'm on Julia 1.1, HTTP 0.8.0#master, Ubuntu 18.04.
```julia
using HTTP
using Sockets

function serve(text::String)
    req_handler = HTTP.RequestHandlerFunction(req -> (println(text); HTTP.Response(200, text)))

    server = Sockets.listen(8000)
    listener = @async HTTP.listen(ip"127.0.0.1", server=server) do http::HTTP.Stream
        HTTP.handle(req_handler, http)
    end

    # wait until user issues a CTRL+C command.
    try while true sleep(0.1) end
    catch err
        if isa(err, InterruptException)
            schedule(listener, InterruptException, error=true) # does not help, so could be omitted
            close(server)
            println("\n✓ Server shut down.")
        else
            throw(err)
        end
    end
    return nothing
end
```
In a fresh Julia session, run the above code, then
```julia
serve("hi")
```
then hit Ctrl-C, go to `localhost:8000` with a browser, tells you it's unable to connect. Fine.

Run `serve("hi-2")`, visit `localhost:8000`, shows you "hi-2" (in browser and Julia console) fine.

Ctrl-C, now Julia is "free" again and you can do other stuff. But visit `localhost:8000` again! It still responds the same, and shows in the Julia console as well!

Then of course, when starting the server with different parameters, say `serve("hi-3")`, this won't change a thing, you'll still get "hi-2" in browser and console.

**Comment by asprionj** [comment]
Ok, probably close to minimal example now. Run:
```julia
using HTTP
using Sockets

server = Sockets.listen(8000)
listener = @async HTTP.listen(ip"127.0.0.1", server=server) do http::HTTP.Stream
    HTTP.handle(HTTP.RequestHandlerFunction(req -> HTTP.Response(200, "hi")), http)
end
```
visit `localhost:8000` with browser, then
```julia
close(server)
```
in Julia. The `listener` task will be in a `:failed` state and you can still visit `localhost:8000` and get a "hi".

**Comment by tlienart** [comment]
Nice yeah that's much simpler!

And for what it's worth, I tried with the #406 and it disconnects as expected.

**Comment by quinnj** [comment]
Fixed via #419
